ili/inference/runner_sbi.py

```diff
@@ -120,33 +125,59 @@ def __call__(self, loader):
 
         Args:
             loader (BaseLoader): data loader with stored summary-parameter pairs
+            or has ability to simulate summary-parameter pairs
+
+        Raises:
+            Exception: cannot perform multi-round inference if loader cannot simulate
         """
 
         t0 = time.time()
-        x = torch.Tensor(loader.get_all_data())
-        theta = torch.Tensor(loader.get_all_parameters())
-        posteriors, val_loss = [], []
+        if hasattr(loader, 'simulate'):
+            theta, x = loader.simulate(self.prior)
+            x_obs= loader.get_obs_data()
+        else:
+            x = torch.Tensor(loader.get_all_data())
+            theta = torch.Tensor(loader.get_all_parameters())
+            if self.num_rounds != 1:
+                raise Exception("If not simulating data you cannot have multi-round inference")
+        all_model = []
         for n, posterior in enumerate(self.neural_posteriors):
+            all_model.append(self.inference_class(
+                    prior=self.prior,
+                    device=self.device,
+                ))
+        proposal = self.prior
+        for rnd in range(self.num_rounds):
+            t1 = time.time()
             logging.info(
-                f"Training model {n+1} out of {len(self.neural_posteriors)} ensemble models"
-            )
-            model = self.inference_class(
-                prior=self.prior,
-                density_estimator=posterior,
-                device=self.device,
+                f"Running round {rnd+1} of {self.num_rounds}"
             )
-            model = model.append_simulations(theta, x)
-            if not isinstance(self.embedding_net, nn.Identity):
-                self.embedding_net.initalize_model(n_input=x.shape[-1])
-            density_estimator = model.train(
-                **self.train_args,
+            if rnd > 0:
+                theta, x = loader.simulate(proposal)
+            posteriors, val_loss = [], []
+            for n, posterior in enumerate(self.neural_posteriors):
+                logging.info(
+                    f"Training model {n+1} out of {len(self.neural_posteriors)} ensemble models"
+                )
+                if not isinstance(self.embedding_net, nn.Identity):
+                    self.embedding_net.initalize_model(n_input=x.shape[-1])
+                density_estimator = all_model[n].append_simulations(theta, x, proposal).train(
```

Comment:
I checked and this does, by default, indeed always train the existing internal neural network every time we call train, without reinitialization ([link](https://github.com/mackelab/sbi/blob/5768a7c05aec799fe435167f75acaca483e5404d/sbi/inference/snpe/snpe_base.py#L304)). It also uses all training data that has been appended, so we only need pass the round-by-round theta's and x's. 

If you wanted a fresh neural network, you'd pass retrain_from_scratch=True. If you wanted to only use this round's data, you'd pass discard_prior_samples=True. Both of these are default = False.

Nothing to be done here, I just wanted to record this for future reference. :)